### ISSUE TITLE
Regression from zero out racily-clean entries

### ISSUE THREAD

**Issue by swisspol** [libgit2/libgit2#3231]
I have one of my high-level functional tests failing due to #3219. Even if in Obj-C, it should be pretty easy to understand what the test does:
```objc
- (void)testSubmoduleEdgeCases {
  NSString* path = [NSTemporaryDirectory() stringByAppendingPathComponent:[[NSProcessInfo processInfo] globallyUniqueString]];
  NSString* output;
  
  // Create submodule repo
  GCRepository* repo = [self createLocalRepositoryAtPath:path bare:NO];
  XCTAssertNotNil(repo);
  XCTAssertNotNil([repo createMockCommitHierarchyFromNotation:@"m0 - m1 - m2<master>" force:NO error:NULL]);
  
  // Add submodule and commit
  output = [self runGitCLTWithRepository:self.repository command:@"submodule", @"add", path, @"repo", nil];
  XCTAssertNotNil(output);
  output = [self runGitCLTWithRepository:self.repository command:@"submodule", @"update", @"--init", nil];
  XCTAssertNotNil(output);
  GCCommit* commit1 = [self.repository createCommitFromHEADWithMessage:@"Added submodule" error:NULL];
  XCTAssertNotNil(commit1);
  
  // Checkout commit before submodule was added
  XCTAssertTrue([self.repository checkoutCommit:self.initialCommit options:kGCCheckoutOption_UpdateSubmodulesRecursively error:NULL]);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqualObjects([self.repository listSubmodules:NULL], @[]);
  
[1]

  // Checkout commit when submodule was added
  XCTAssertTrue([self.repository checkoutLocalBranch:[self.repository findLocalBranchWithName:@"master" error:NULL] options:kGCCheckoutOption_UpdateSubmodulesRecursively error:NULL]);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqual([[self.repository listSubmodules:NULL] count], 1);
  
  // Delete submodule and commit
  output = [self runGitCLTWithRepository:self.repository command:@"submodule", @"deinit", @"repo", nil];
[2]  XCTAssertNotNil(output);
  output = [self runGitCLTWithRepository:self.repository command:@"rm", @"repo", nil];
  XCTAssertNotNil(output);
  GCCommit* commit2 = [self.repository createCommitFromHEADWithMessage:@"Removed submodule" error:NULL];
  XCTAssertNotNil(commit2);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqualObjects([self.repository listSubmodules:NULL], @[]);
  
  // Checkout commit when submodule was added
  XCTAssertTrue([self.repository checkoutCommit:commit1 options:kGCCheckoutOption_UpdateSubmodulesRecursively error:NULL]);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqual([[self.repository listSubmodules:NULL] count], 1);
  
  // Checkout commit when submodule was deleted
  XCTAssertTrue([self.repository checkoutLocalBranch:[self.repository findLocalBranchWithName:@"master" error:NULL] options:kGCCheckoutOption_UpdateSubmodulesRecursively error:NULL]);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqualObjects([self.repository listSubmodules:NULL], @[]);
  
  // Checkin file where submodule was
  [self makeCommitWithUpdatedFileAtPath:@"repo" string:@"nothing" message:@"NOTHING"];
  
  // Checkout commit when submodule was added
  XCTAssertTrue([self.repository checkoutCommit:commit1 options:kGCCheckoutOption_UpdateSubmodulesRecursively error:NULL]);
  XCTAssertFalse([self.repository checkRepositoryDirty:YES]);
  XCTAssertEqual([[self.repository listSubmodules:NULL] count], 1);
  
  // Destroy submodule repo
  [self destroyLocalRepository:repo];
}
```

[2] is where the failure happens: calling `git submodule deinit repo` fails with:
```
fatal: Please, stage your changes to .gitmodules or stash them to proceed
Submodule work tree 'repo' contains local modifications; use '-f' to discard them
```

However there are actually no changes in the repo. What's happening is that libgit2 has written an index file that reflects changes that don't actually exist on disk. If you call `git status` before calling `git submodule deinit repo`, which refreshes the index, then it works.

If you add a `sleep(1);` call at [1], it also works.

**Comment by swisspol** [comment]
Reverting 316b820b6fed712d4b6fae886ba1e6c44afc6c81 fixes the issue.

**Comment by carlosmn** [comment]
I would be surprised if zeroing out the size causes a regression. It is much more likely that it's exposing race conditions or other bugs which were being hidden before. Each of the racy-git PRs has shown that our merge tests behaves incorrectly in some cases.

**Comment by swisspol** [comment]
If you look at the test above, you will notice it actually only use libgit2 to do the checkouts, while before and after operations are done using Git CLT. I doubt that's exposing some hidden bugs in libgit2. Seems more like this is libgit2 generating an index file that confuses Git CLT.

After doing more testing, it looks like the bug is related to `.gitmodules`. If you do 2 checkouts very fast that update this file, that triggers the problem. I can't repro the bug by doing 2 checkouts that only update other "regular" files.

Furthermore doing the exact same 2 checkouts in Git CLT doesn't trigger a similar issue:
```sh
git checkout HEAD~1 && git checkout master && git submodule deinit repo
```

While this sequences triggers the bug:
```
(libgit2) checkout HEAD~1 && (libgit2) checkout master && git submodule deinit repo
```

Are you sure Git core is not special-casing submodules when handling racy-git?

**Comment by swisspol** [comment]
Another workaround: if you call `git_diff_index_to_workdir()` with `GIT_DIFF_UPDATE_INDEX` after the 2 checkouts but before `git submodule deinit repo`, it will update and fix the index.

**Comment by swisspol** [comment]
Another strange thing: if you call `git_diff_index_to_workdir()` with `GIT_DIFF_UPDATE_INDEX` after the 2 checkouts but before `git submodule deinit repo`, it will correctly detect that `.gitmodules` is out of sync, and end up rewriting the index to disk, but `git submodule deinit repo` still fails with
```
fatal: Please, stage your changes to .gitmodules or stash them to proceed
Submodule work tree 'repo' contains local modifications; use '-f' to discard them
```

**Comment by carlosmn** [comment]
It does indeed ignore all of the racy smudging for submodules. There's also this extra stat which we'd have to do which is missing. I'll update this in the pending PR.

**Comment by swisspol** [comment]
You know, I have by now a good set of functional tests for my libgit2 wrapper, and if you CC me on "risky" incoming PRs, I can likely pre-flight them with these tests and point out any edge cases. It has been a few times that I find such edge cases after a PR has been merged.

**Comment by ethomson** [comment]
>  It has been a few times that I find such edge cases after a PR has been merged.

Indeed it has.  And certainly we appreciate you pointing out bugs or possible issues, but what would be *amazing* is if you would contribute unit tests for libgit2 here.  I don't think pinging every user of libgit2 every time we merge a PR is going to scale well.

**Comment by carlosmn** [comment]
The fix for the submodules, among others, was merged with #3226
